src/analysis/processing/qgsalgorithmlinesubstring.cpp

```diff
@@ -0,0 +1,163 @@
+  startDistance->setDynamicLayerParameterName( QStringLiteral( "INPUT" ) );
+  addParameter( startDistance.release() );
+
+  std::unique_ptr< QgsProcessingParameterDistance> endDistance = qgis::make_unique< QgsProcessingParameterDistance >( QStringLiteral( "END_DISTANCE" ),
+      QObject::tr( "End distance" ), 1.0, QStringLiteral( "INPUT" ), false, 0 );
+  endDistance->setIsDynamic( true );
+  endDistance->setDynamicPropertyDefinition( QgsPropertyDefinition( QStringLiteral( "End Distance" ), QObject::tr( "End distance" ), QgsPropertyDefinition::DoublePositive ) );
+  endDistance->setDynamicLayerParameterName( QStringLiteral( "INPUT" ) );
+  addParameter( endDistance.release() );
+}
+
+QgsProcessingFeatureSource::Flag QgsLineSubstringAlgorithm::sourceFlags() const
+{
+  // skip geometry checks - this algorithm doesn't care about invalid geometries
+  return QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks;
+}
+
+bool QgsLineSubstringAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mStartDistance = parameterAsDouble( parameters, QStringLiteral( "START_DISTANCE" ), context );
+  mDynamicStartDistance = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "START_DISTANCE" ) );
+  if ( mDynamicStartDistance )
+    mStartDistanceProperty = parameters.value( QStringLiteral( "START_DISTANCE" ) ).value< QgsProperty >();
+
+  mEndDistance = parameterAsDouble( parameters, QStringLiteral( "END_DISTANCE" ), context );
+  mDynamicEndDistance = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "END_DISTANCE" ) );
+  if ( mDynamicEndDistance )
+    mEndDistanceProperty = parameters.value( QStringLiteral( "END_DISTANCE" ) ).value< QgsProperty >();
+
+  return true;
+}
+
+QgsFeatureList QgsLineSubstringAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsFeature f = feature;
+  if ( f.hasGeometry() )
+  {
+    const QgsGeometry geometry = f.geometry();
+    double startDistance = mStartDistance;
+    if ( mDynamicStartDistance )
+      startDistance = mStartDistanceProperty.valueAsDouble( context.expressionContext(), startDistance );
+
+    double endDistance = mEndDistance;
+    if ( mDynamicEndDistance )
+      endDistance = mEndDistanceProperty.valueAsDouble( context.expressionContext(), endDistance );
+
+    const QgsCurve *curve = nullptr;
+    if ( !geometry.isMultipart() )
+      curve = qgsgeometry_cast< const QgsCurve * >( geometry.constGet() );
+    else
+    {
+      if ( const QgsGeometryCollection *collection = qgsgeometry_cast< const QgsGeometryCollection * >( geometry.constGet() ) )
+      {
+        if ( collection->numGeometries() > 0 )
+        {
+          curve = qgsgeometry_cast< const QgsCurve * >( collection->geometryN( 0 ) );
```

Comment:
What do you think about adding a note in the help string for this particular case?

Reply:
Actually, I should handle this case properly - I'll upgrade the algorithm to do that.

Reply:
Looking into this, it's non trivial to implement and I suspect the results will never be as intended - I'll just make a note of the current behaviour in the docs instead.